frappe_s3_attachment/controller.py

```diff
@@ -292,6 +305,91 @@ def upload_existing_files_s3(name, file_name):
     else:
         pass
 
+# download s3 files
+def download_s3_files(private_local_folder_path, public_local_folder_path):
+    s3_download = S3Operations()
+    bucket = s3_download.S3_RESOURCE.Bucket(s3_download.BUCKET)
+    for obj in bucket.objects.filter(Prefix=s3_download.folder_name):
+        # Check if the object is a file and not a folder
+        if not obj.key.endswith('/'):
+            download_s3_file(obj, s3_download.BUCKET, private_local_folder_path, public_local_folder_path)
+
+# download s3 file
+def download_s3_file(obj, bucket_name, private_local_folder_path, public_local_folder_path):
+    # check from file_url that file is in s3 or local and ommit download if it is in local
+    name = frappe.db.sql(f"""select `name` from `tabFile` where `file_url` LIKE '%{obj.key}%'""")
```

Comment:
There's a function in original code that matches the file name with URL format. You must use the same function to achieve idempotency instead of using your own.

Reply:
This function will get unique index name from database.
Here, obj.key will also be unique.
The function who matches file url is regex function in controller.py in frappe-s3
It will return only true or false.